### PR TITLE
Variance fix

### PR DESCRIPTION
**Pull request by typerSniper**
Type parameters of `iset` and domain of `imap` are now checked with a `Covariant_Permissive` context. If a parameter specified to be strict is used in those positions then dafny will give an error. Concrete examples are written in `Variance.dfy`.

### CODE REVIEW

**Comment by robin-aws** [comment]
The failing checks are broken on master too - this PR fixes them: https://github.com/dafny-lang/dafny/pull/1237

File: Source/Dafny/Resolver.cs [hunk 1]
```diff
@@ -14140,28 +14140,36 @@ class LazyString_OnTypeEquals
...
-    /// * "leftOfArrow" says that the context is to the left of some arrow
+    /// * "lax" says that the context is not strict -- type parameters declared to be strict must not be used in a lax context
     /// </summary>
-    public void CheckVariance(Type type, ICallable enclosingTypeDefinition, TypeParameter.TPVariance context, bool leftOfArrow) {
+    public void CheckVariance(Type type, ICallable enclosingTypeDefinition, TypeParameter.TPVariance context, bool lax) {
       Contract.Requires(type != null);
       Contract.Requires(enclosingTypeDefinition != null);
 
       type = type.Normalize();  // we keep constraints, since subset types have their own type-parameter variance specifications; we also keep synonys, since that gives rise to better error messages
+
```
Comment (RustanLeino) [inline Source/Dafny/Resolver.cs:14150]: This and L14187 below make me think your IDE is set to insert blank lines before `if` statements or blocks. If you didn't intentionally insert this line, please adjust your IDE style formatting settings.

File: Source/Dafny/Resolver.cs [hunk 2]
```diff
@@ -14140,28 +14140,36 @@ class LazyString_OnTypeEquals
...
       if (type is BasicType) {
         // fine
       } else if (type is MapType) {
         var t = (MapType)type;
-        CheckVariance(t.Domain, enclosingTypeDefinition, context, leftOfArrow);
-        CheckVariance(t.Range, enclosingTypeDefinition, context, leftOfArrow);
+        // If its an infinite map, the domain's context is lax
+        CheckVariance(t.Domain, enclosingTypeDefinition, context,
+          lax || !t.Finite);
```
Comment (RustanLeino) [inline Source/Dafny/Resolver.cs:14157]: Make this one line. (If the change was done by the IDE formatting, you can change the line width to something much larger. I think even 150 characters is reasonable with today's wide screens.)
Same comment on L14163.

File: Source/Dafny/Resolver.cs [hunk 3]
```diff
@@ -14140,28 +14140,36 @@ class LazyString_OnTypeEquals
...
+        CheckVariance(t.Range, enclosingTypeDefinition, context, lax);
+      } else if(type is SetType){
```
Comment (RustanLeino) [inline Source/Dafny/Resolver.cs:14159]:
```suggestion
      } else if (type is SetType) {
```

**Review by RustanLeino — CHANGES_REQUESTED** [review]
I have some minor formatting comments. Please address.
Thanks for adding the many new good new tests. Please also add the specific example from https://github.com/dafny-lang/dafny/issues/1219.

**Review by RustanLeino — APPROVED** [review]

**Pull request merged**